### ISSUE TITLE
Flake: `iptables: create-instance-chains: iptables: Memory allocation problem.`

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#3127]
## Flake Report
Occasionally it seems like Garden fails [here](https://github.com/cloudfoundry/guardian/blob/ac7db3af3e59ae48f4f9c1f0c2e7074cf94e35e2/kawasaki/iptables/iptables.go#L64-L66
) when Concourse is trying to create a task container. 

## Examples
 - https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/testflight/builds/382
 - PCI team in the Toronto office saw this in their Concourse ( 4.2.1 ) environment as well

## Observations
  - The worker metrics don't show any abnormalities in terms of memory utilization https://snapshot.raintank.io/dashboard/snapshot/2LRam0LlWhwwEMP38tr6MLNZ4FV6gaAL?orgId=2

## Additional Context
@kcmannem has a stack dump from PCI's issue

## Version Info

* Concourse version: 4.2.1 & 5.0.0
* Deployment type (BOSH/Docker/binary): BOSH
* Infrastructure/IaaS: GCP
* Did this used to work? 🤷‍♂️

**Comment by kcmannem** [comment]
Garden team investigated this issue. They said just increase the workers memory. There is no other fix they can provide.

**Comment by cirocosta** [comment]
Which seems like a very weird remediation in this case - the workers were all at 20% utilization in our case (w/out limits being provided to the containers) :thinking:

**Comment by vito** [comment]
Yeah...gotta say I'm pretty skeptical. Our workers never really seem to show high memory usage. They already have 16GB each, I don't think this is a situation where we can just throw hardware at it.

The only scenario I can imagine is a very quick spike in memory allocation that goes down just as quickly, but I find it hard to imagine a spike big enough to allocate ~12GB of RAM, let alone one that's quick enough to escape our metrics.

**Comment by kcmannem** [comment]
Heres what they had to say:

_"A hypothesis we had was that since this is only observed on iptables operations that perhaps the table within kernel memory keeping track of this stuff filled up. I’ve heard from people that iptable is pretty inefficient with memory when it’s persisting rules."_

**Comment by vito** [comment]
One thing of note is that build runs containers in Docker Compose, in a container. Could `iptables` rules be sticking around between builds? Or do they get cleaned up properly when the network namespace is destroyed?

**Comment by topherbullock** [comment]
Yeah, I don't think more worker memory will help here; the workers' memory utilization ~15% when the flake occurred. 

@vito Garden _should_ destroy the `iptables` for the container when the container is destroyed (https://github.com/cloudfoundry/guardian/blob/7536c003d476be01a24401148db6238416f53c02/gardener/gardener_test.go#L830-L835, https://github.com/cloudfoundry/guardian/blob/7536c003d476be01a24401148db6238416f53c02/kawasaki/networker.go#L246-L278). I think any `iptables` entries created by Docker will go away; we'd probably notice a lot more issues if they were leaking

**Comment by davidjahn** [comment]
Just chiming in that we are noticing this flake as well when we are running slightly more containers/worker than usual (~160 on one with highest load).

**Comment by marco-m** [comment]
We started seeing the same with 5.0 RC54

**Comment by cirocosta** [comment]
Hey @marco-m ,

Do you also see this happening only in workers with more containers than the usual like @davidjahn ?

thx!

**Comment by wagdav** [comment]
@cirocosta 

@marco-m  and I summarized what we see in #3251

**Comment by edude03** [comment]
We're seeing this with 5.0.0-rc4 when there are more builds than usual as well.

**Comment by marco-m** [comment]
To summarize: we are currently using RC78, which contains the fix for #3251. We don't see any more the `iptables` error. We have a distribution of containers per worker of average 100 (very low because we are currently running overprovisioned).

@edude03 I strongly suggest to upgrade and to try `fewest` as container placement srategy; RC4 was pre #3251.

**Comment by lrstanley** [comment]
Noticed this today as well.. still running v3.14.1, 16 workers in the environment this happened in, most at about ~120-150, but we cap at 250. I don't think it's necessarily due to the amount of containers (as, we frequently have workers peak over 200 without issue). I think I've only seen it a few times, so hard to say. Def. not seeing any memory issues, disk or inode issues, etc. We currently use random placement strategy as it's not a thing in the version we're running (upgrade is a work in progress.. a bit hard to do with almost 1,900 pipelines and ~240 teams).

**Comment by cirocosta** [comment]
Hey,

I was giving a deeper look into this, and it turns out that everytime such problem occurs, we can get a stack trace from `dmesg`:

```
[2559241.728047] warn_alloc: 154 callbacks suppressed
[2559241.728049] iptables: vmalloc: allocation failure, allocated 0 of 475136 bytes, mode:0x14010c0(GFP_KERNEL|__GFP_NORETRY), nodemask=(null)
[2559241.745870] iptables cpuset=8b477909a087561bc752689fd4c91329e4e68da6d6692321249b6182ab68bd1d mems_allowed=0
[2559241.755981] CPU: 2 PID: 972000 Comm: iptables Tainted: G        W       4.14.89+ #1
[2559241.763934] Hardware name: Google Google Compute Engine/Google Compute Engine, BIOS Google 01/01/2011
[2559241.773464] Call Trace:
[2559241.776222]  dump_stack+0x63/0x82
[2559241.779837]  warn_alloc+0xe8/0x190
[2559241.783533]  __vmalloc_node_range+0x1b4/0x260
[2559241.788188]  __vmalloc_node_flags_caller+0x39/0x40
[2559241.793295]  ? xt_alloc_table_info+0x41/0x90
[2559241.801695]  xt_alloc_table_info+0x41/0x90
[2559241.806098]  do_ipt_set_ctl+0xd7/0x1b0
[2559241.810153]  nf_setsockopt+0x4a/0x70
[2559241.814033]  SyS_setsockopt+0xc5/0x140
[2559241.818085]  do_syscall_64+0x70/0x200
[2559241.822051]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[2559241.827411] RIP: 0033:0x7fc1bef40e6a
[2559241.831290] RSP: 002b:00007ffe19576ab8 EFLAGS: 00000206 ORIG_RAX: 0000000000000036
[2559241.839158] RAX: ffffffffffffffda RBX: 00000000007a7970 RCX: 00007fc1bef40e6a
[2559241.846601] RDX: 0000000000000040 RSI: 0000000000000000 RDI: 0000000000000004
[2559241.854039] RBP: 00007fc1bf6e7070 R08: 0000000000072c88 R09: 00007fc1bf759860
[2559241.861478] R10: 00007fc1bf6e7010 R11: 0000000000000206 R12: 00000000007a7978
[2559241.868909] R13: 00000000007a7978 R14: 00007fc1bf6e7010 R15: 0000000000000000
[2559241.876387] Mem-Info:
[2559241.878981] active_anon:411850 inactive_anon:44802 isolated_anon:0
                  active_file:320486 inactive_file:5926845 isolated_file:0
                  unevictable:0 dirty:156362 writeback:113 unstable:0
                  slab_reclaimable:913714 slab_unreclaimable:221869
                  mapped:62493 shmem:45001 pagetables:13279 bounce:0
                  free:49961 free_pcp:2310 free_cma:0
[2559241.915316] Node 0 active_anon:1647604kB inactive_anon:179208kB active_file:1281308kB inactive_file:23705972kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:249948kB dirty:611680kB writeback:4632kB shmem:180004kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 0kB writeback_tmp:0kB unstable:0kB all_unreclaimable? no
[2559241.944755] Node 0 DMA free:15908kB min:32kB low:44kB high:56kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15992kB managed:15908kB mlocked:0kB kernel_stack:0kB pagetables:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
[2559241.970809] lowmem_reserve[]: 0 2991 32153 32153
[2559241.975771] Node 0 DMA32 free:121464kB min:6284kB low:9344kB high:12404kB active_anon:61004kB inactive_anon:3120kB active_file:22624kB inactive_file:1154272kB unevictable:0kB writepending:159880kB present:3129328kB managed:3063760kB mlocked:0kB kernel_stack:1972kB pagetables:1000kB bounce:0kB free_pcp:2684kB local_pcp:4kB free_cma:0kB
[2559242.005833] lowmem_reserve[]: 0 0 29161 29161
[2559242.010523] Node 0 Normal free:51300kB min:61264kB low:91124kB high:120984kB active_anon:1586972kB inactive_anon:176088kB active_file:1257944kB inactive_file:22562060kB unevictable:0kB writepending:445516kB present:30408704kB managed:29864764kB mlocked:0kB kernel_stack:96524kB pagetables:52116kB bounce:0kB free_pcp:7884kB local_pcp:116kB free_cma:0kB
[2559242.042159] lowmem_reserve[]: 0 0 0 0
[2559242.046157] Node 0 DMA: 1*4kB (U) 0*8kB 0*16kB 1*32kB (U) 2*64kB (U) 1*128kB (U) 1*256kB (U) 0*512kB 1*1024kB (U) 1*2048kB (M) 3*4096kB (M) = 15908kB
[2559242.059895] Node 0 DMA32: 31*4kB (UME) 91*8kB (UME) 1336*16kB (UE) 2788*32kB (UME) 124*64kB (UME) 0*128kB 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 119380kB
[2559242.074587] Node 0 Normal: 13422*4kB (UME) 11*8kB (M) 12*16kB (M) 0*32kB 0*64kB 0*128kB 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 53968kB
[2559242.088621] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[2559242.097543] 6288712 total pagecache pages
[2559242.101892] 0 pages in swap cache
[2559242.105541] Swap cache stats: add 0, delete 0, find 0/0
[2559242.111105] Free swap  = 0kB
[2559242.114344] Total swap = 0kB
[2559242.117666] 8388506 pages RAM
[2559242.120959] 0 pages HighMem/MovableOnly
[2559242.125120] 152398 pages reserved
```

Looking for possible reports for this, I've found https://bugzilla.kernel.org/show_bug.cgi?id=200651, which presents a reproducible that seems quite related to our case.

In that reproducible, the author is literally, at the same time:
- creating a bunch of cgroups (which is something that we do when we create a bunch of containers),
- exercising some memory pressure (which seems reasonably on point - we usually see this problem when there's a bunch of steps happening at the same time), and
- performing an `iptables-restore` (well, might seem a bit unrelated but the similarity here is that when we're creating containers, we're also setting up networking, which involves a bunch of iptables - see Guardian's Kawasaki).

From what I understand (very little), the problem arises from the fact that the allocations that the kernel-space implementation of iptables didn't use the right "get free page" (GFP) flag.

The fix is present in kernels >= v4.19-rc1: https://github.com/torvalds/linux/commit/a148ce15375fc664ad64762c751c0c2aecb2cafe

As next steps, we could:

1. try reaching a consistent reproducible using pure container creation + memory pressure, then
2. try such reproducible on v.19-rc1+  (which the recently released Ubuntu Disco has), then, if all it confirmed,
3. starting some discussions with canonical about a possible backport to Xenial.

On a totally different path, I remember @kcmannem told me about potentially using an even more lightweight external network plugin for `garden-runc` that'd completely bypass iptables by leveraging `tc + ebpf`. Have you worked on that, @kcmannem ?

Thanks!

**Comment by kcmannem** [comment]
@cirocosta nope but more reason to investigate it

**Comment by danail-branekov** [comment]
Hi Concourse friends,

While investigating probably unrelated thing in Garden we managed to reproduce consistently that issue and maybe you would be interested in the scenario.

We created an AWS concourse deployment (if version matters, we used v.4.2.2) with one worker on a t3.nano VM (0.5G of memory) - we really wanted to put memory pressure on that worker. Then we created the following dummy task:
```yml
---
image_resource:
  type: docker-image
  source:
    repository: cfgarden/garden-ci
    tag: '2.0.11'

platform: linux

run:
  path: /bin/echo
```

Then we ran 8 parallel infinite loops running the same task over and over again. 
```bash
for i in `seq 0 200` ; do echo Attempt $i && fly --target slab-ci e -p -c "<the task yml>"; done
```

We reproduced the issue quite quickly. After the issue has been reproduced the worker seems to become unusable and needs to be recreated. On such a broken worker VM we observed tons of `gdn` and `baggegeclaim` processes running even though we have interrupted all the instances of infinite loops. Killing those processes does not recover the worker, as I said we had to recreate the VM

Note that our endless loop runs fine when run in 4 instances

**Comment by kcmannem** [comment]
@danail-branekov thanks for the steps. We'll be sure to test this out more.

**Comment by cdutra** [comment]
The Release Integration started experiencing this as well when we upgraded to Concourse `5.1.0`.

**Comment by marco-m** [comment]
@cirocosta 
> tl;dr: it seems like [torvalds/linux@a148ce1]
(https://github.com/torvalds/linux/commit/a148ce15375fc664ad64762c751c0c2aecb2cafe) (present in Linux 4.19+ is the fix for this problem).

Did you try a kernel recent enough to contain that commit? Do you plan to try it?

**Comment by ddadlani** [comment]
We want to validate that this does not occur anymore in v5+ with the patch, in which case we might want to ask Canonical to backport it.

**Comment by eedwards-sk** [comment]
Hey Y'all

I reproduced this issue in Concourse 5.3.0 with Ubuntu Bionic (18.04)

**Comment by cirocosta** [comment]
> Did you try a kernel recent enough to contain that commit? Do you plan to try it?


I had tried in a very non-scientific way - having thousands of containers landing in a single worker 😅  Couldn't reproduce the issue in that one (Ubuntu Disco - which ships w/ linux 5+)

**Comment by colmtuohy** [comment]
I saw the issue temporally on Concourse 5.2.0 for the email put resource (iptables: create-instance-chains: iptables: Memory allocation problem). Not sure what cleared the issue

**Comment by Rukenshia** [comment]
We just upgraded to concourse 5.3.0 and also experience this problem.

We use this stemcell in bosh:

```
- alias: xenial
  os: ubuntu-xenial
  version: 315.64
```

**Comment by suhlig** [comment]
Ditto with 5.4.0 using the binaries on Ubuntu 18.04:

```
# uname -a
Linux ci 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic
```

**Comment by ddadlani** [comment]
we're going to push this investigation to after our containerd spike since this might go away if we move to containerd. For now the only lead we have is that it might need a kernel version bump.

**Comment by jwatson0** [comment]
Got this today.  Here's my dmesg stack trace in case it helps:
```
[4595855.660122] iptables: vmalloc: allocation failure, allocated 0 of 331776 bytes, mode:0x14010c0(GFP_KERNEL|__GFP_NORETRY), nodemask=(null)
[4595855.660124] iptables cpuset=/ mems_allowed=0
[4595855.660129] CPU: 1 PID: 13184 Comm: iptables Not tainted 4.15.0-1044-aws #46-Ubuntu
[4595855.660130] Hardware name: Amazon EC2 c5d.large/, BIOS 1.0 10/16/2017
[4595855.660131] Call Trace:
[4595855.660137]  dump_stack+0x63/0x8b
[4595855.660140]  warn_alloc+0xff/0x1a0
[4595855.660144]  __vmalloc_node_range+0x170/0x270
[4595855.660145]  __vmalloc_node_flags_caller+0x47/0x50
[4595855.660151]  ? xt_alloc_table_info+0x39/0x80 [x_tables]
[4595855.660154]  kvmalloc_node+0x61/0x80
[4595855.660156]  xt_alloc_table_info+0x39/0x80 [x_tables]
[4595855.660159]  do_ipt_set_ctl+0xd0/0x1c0 [ip_tables]
[4595855.660163]  nf_setsockopt+0x4c/0x80
[4595855.660165]  ip_setsockopt+0x6c/0x80
[4595855.660169]  raw_setsockopt+0x34/0x40
[4595855.660172]  sock_common_setsockopt+0x1a/0x20
[4595855.660174]  SyS_setsockopt+0x86/0xf0
[4595855.660178]  do_syscall_64+0x73/0x130
[4595855.660181]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[4595855.660183] RIP: 0033:0x7f7eacd56e6a
[4595855.660184] RSP: 002b:00007fff77e1d598 EFLAGS: 00000202 ORIG_RAX: 0000000000000036
[4595855.660185] RAX: ffffffffffffffda RBX: 0000000000bc9970 RCX: 00007f7eacd56e6a
[4595855.660186] RDX: 0000000000000040 RSI: 0000000000000000 RDI: 0000000000000004
[4595855.660187] RBP: 00007f7ead53e070 R08: 000000000004ff48 R09: 00007f7ead58dc60
[4595855.660188] R10: 00007f7ead53e010 R11: 0000000000000202 R12: 0000000000bc9978
[4595855.660189] R13: 0000000000bc9978 R14: 00007f7ead53e010 R15: 0000000000000000
[4595855.660190] Mem-Info:
[4595855.660194] active_anon:182253 inactive_anon:31 isolated_anon:0
                  active_file:239432 inactive_file:367465 isolated_file:0
                  unevictable:0 dirty:2280 writeback:130 unstable:0
                  slab_reclaimable:46719 slab_unreclaimable:65974
                  mapped:21603 shmem:1290 pagetables:4703 bounce:0
                  free:21165 free_pcp:65 free_cma:0
[4595855.660197] Node 0 active_anon:729012kB inactive_anon:124kB active_file:957728kB inactive_file:1469860kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:86412kB dirty:9120kB writeback:520kB shmem:5160kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 0kB writeback_tmp:0kB unstable:0kB all_unreclaimable? no
[4595855.660197] Node 0 DMA free:14920kB min:284kB low:352kB high:420kB active_anon:676kB inactive_anon:0kB active_file:0kB inactive_file:288kB unevictable:0kB writepending:52kB present:15992kB managed:15908kB mlocked:0kB kernel_stack:0kB pagetables:4kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
[4595855.660201] lowmem_reserve[]: 0 2972 3666 3666 3666
[4595855.660203] Node 0 DMA32 free:57048kB min:54568kB low:68208kB high:81848kB active_anon:482000kB inactive_anon:48kB active_file:826444kB inactive_file:1291672kB unevictable:0kB writepending:6928kB present:3129320kB managed:3063752kB mlocked:0kB kernel_stack:17400kB pagetables:11032kB bounce:0kB free_pcp:68kB local_pcp:52kB free_cma:0kB
[4595855.660207] lowmem_reserve[]: 0 0 693 693 693
[4595855.660209] Node 0 Normal free:12692kB min:12728kB low:15908kB high:19088kB active_anon:246336kB inactive_anon:76kB active_file:131916kB inactive_file:177796kB unevictable:0kB writepending:2564kB present:800768kB managed:714624kB mlocked:0kB kernel_stack:6264kB pagetables:7776kB bounce:0kB free_pcp:192kB local_pcp:192kB free_cma:0kB
[4595855.660212] lowmem_reserve[]: 0 0 0 0 0
[4595855.660215] Node 0 DMA: 4*4kB (UMH) 3*8kB (UME) 4*16kB (UME) 3*32kB (MEH) 4*64kB (UEH) 3*128kB (UEH) 3*256kB (UEH) 2*512kB (UE) 4*1024kB (UMEH) 2*2048kB (UE) 1*4096kB (M) = 14920kB
[4595855.660225] Node 0 DMA32: 7421*4kB (UMEH) 1624*8kB (UMEH) 667*16kB (UMEH) 87*32kB (UMEH) 5*64kB (MH) 6*128kB (H) 1*256kB (H) 0*512kB 0*1024kB 0*2048kB 0*4096kB = 57476kB
[4595855.660233] Node 0 Normal: 2798*4kB (UMEH) 302*8kB (UMEH) 0*16kB 0*32kB 0*64kB 0*128kB 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 13608kB
[4595855.660241] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=1048576kB
[4595855.660242] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[4595855.660243] 602190 total pagecache pages
[4595855.660244] 0 pages in swap cache
[4595855.660245] Swap cache stats: add 0, delete 0, find 0/0
[4595855.660245] Free swap  = 0kB
[4595855.660246] Total swap = 0kB
[4595855.660246] 986520 pages RAM
[4595855.660247] 0 pages HighMem/MovableOnly
[4595855.660247] 37949 pages reserved
[4595855.660248] 0 pages cma reserved
[4595855.660248] 0 pages hwpoisoned
```

**Comment by marco-m** [comment]
@jwatson0 which version of Concourse ?

**Comment by matt-royal** [comment]
@suhlig It looks like Ubuntu 18.04.2 doesn't have the version of the kernel with the fix. It ships with kernel 4.18. Ubuntu 18.04.3 ships with kernel 5.0. Let us know if you get a chance to test it with the new kernel.

Unfortunately, my team deploys our Concourses with bosh and there's no short-term plan to support any Ubuntu release newer than Xenial, so there's little hope of us being able to try with kernel 4.19 or newer.

**Comment by mreider** [comment]
Hell friends. We have the same problem as @matt-royal - so we are curious how the Concourse team is deploying this. Clearly they are not using BOSH, for if they were, the stem cell would cause issues. This is causing a fair amount of churn with our pipelines and making product managers (like me) ask why we are using it in the first place.

**Comment by pivotal-jamie-klassen** [comment]
@mreider we're using the [helm chart](https://github.com/concourse/concourse-chart) on k8s (specifically GKE, where node images since `v1.15.4-gke.22` have had the necessary kernel)! That said, we do have some pretty hefty [integration test suites](https://ci.concourse-ci.org/teams/main/pipelines/concourse?group=bosh) that use BOSH using the GCP Xenial stemcells from bosh.io -- maybe we're conveniently missing the scenarios where this bug tends to creep up.

**Comment by blairboy362** [comment]
Hi friends. We hit this recently and I thought I'd share my conclusions in case it helps anyone.

It's a memory fragmentation problem. I'm not familiar enough with concourse to know how or why it got worse recently, but I'm confident concourse is responsible for the majority of the fragmentation. Using the example from @jwatson0's [comment above](https://github.com/concourse/concourse/issues/3127#issuecomment-538581294):

```
[4595855.660122] iptables: vmalloc: allocation failure, allocated 0 of 331776 bytes, mode:0x14010c0(GFP_KERNEL|__GFP_NORETRY), nodemask=(null)
...snip...
[4595855.660233] Node 0 Normal: 2798*4kB (UMEH) 302*8kB (UMEH) 0*16kB 0*32kB 0*64kB 0*128kB 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 13608kB
```

iptables is attempting to allocate 324 KiB and failing. There are no slabs 512 KiB (or bigger; the next size up from the required 324 KiB) available. `cat /proc/buddyinfo` gives the same info as the second line above.

![image](https://user-images.githubusercontent.com/12373655/88391820-d040c580-cdb2-11ea-89fd-7097e8283361.png)
This graph is from our investigation showing the difference in fragmentation between a node running a concourse worker and one that isn't. The metrics come from the prometheus node exporter with the `buddy` collector enabled.

In our context we're running concourse on kubernetes (using the helm chart). As one might imagine, and as others have commented on above, the more stuff running on a single node the bigger the iptables configuration and the more likely this problem is to be hit. We've worked around it so far by forcing the worker pods onto their own nodes. But that will only push the problem downstream a bit.

**Comment by vito** [comment]
@blairboy362 Thanks for sharing! That sounds like a useful thing to watch out for in the future.

Giving worker pods their own nodes makes a lot of sense given the current way things work. In the future I think that will be ultimately resolved with our own K8s native runtime, which we're planning to start very soon (possibly next week).

In the meantime, this may improve with our upcoming `containerd` worker backend which configures fewer rules (https://github.com/concourse/concourse/issues/5534#issuecomment-644862599).

TBH if the real problem is that the worker genuinely cannot allocate memory for some other arbitrary reason, I'm not totally sure what we could do. 🤔 I wonder if using some interface other than `iptables` would help.

Is anyone still running into this issue in a single-dedicated-worker-VM context?
